pkg/obfuscator/replacement_reporter.go

```diff
@@ -0,0 +1,29 @@
+package obfuscator
+
+// this struct mainly exists in case we later want to make it thread-safe, so we don't have to individually go through
+// dozens of obfuscators.
+
+type ReplacementReporter interface {
+	// Report returns a mapping of strings which were replaced by this obfuscator
+	Report() map[string]string
+
+	// UpsertReplacement will add a new replacement along with its original string to the report.
+	// Overwrites existing values for the same key, thus "upsert" for update+insert.
+	UpsertReplacement(original string, replacement string)
+}
+
+type SimpleReporter struct {
+	mapping map[string]string
+}
+
+func (s *SimpleReporter) Report() map[string]string {
+	return s.mapping
+}
+
+func (s *SimpleReporter) UpsertReplacement(original string, replacement string) {
```

Comment:
Is there actually a use-case for upserting? Because I think if it happens then we've done something wrong.

Reply:
depends on the replacementType ultimately, I assume we have to sort this out when we do the consistent replacement algorithm.

Reply:
Even with consistent replacement I don't think we would want to update any existing keys right?

Reply:
who should control that behaviour? is that the "replacer" (which we don't have yet) that ensures it always maps to the same output or the reporter that keeps track of what was replaced? 

I'll change that for the reporting however for additional safety, even though for me that would be a leaking abstraction from the replacer.

Reply:
The way I see it the Obfuscator is the replacer here. For the case of random replacements we should not track what was replaced because that data is quite useless. For consistent/static replacement we should track replacements. And the replacements will always be the same. So if a key exists then it's value should always be used. It should not be overwritten. And it's not leaking abstractions because that's a desired property of the reporter(no dataloss).

Reply:
I added that check as a panic, hope that works for you @arjunrn?

Reply:
> The way I see it the Obfuscator is the replacer here. For the case of random replacements we should not track what was replaced because that data is quite useless. For consistent/static replacement we should track replacements. And the replacements will always be the same. So if a key exists then it's value should always be used. It should not be overwritten. And it's not leaking abstractions because that's a desired property of the reporter(no dataloss).

it's a matter of separation of concerns. The reporter should only report what I pass into it, not control what kind of values are permitted. 

>  For consistent/static replacement we should track replacements.

exactly, but who tracks that replacement? not the reporter I would say.